File: app/src/main/java/cl/figonzal/lastquakechile/services/notifications/FirebaseQuakeNotificationService.java
package cl.figonzal.lastquakechile.services.notifications;

import androidx.annotation.NonNull;

import com.google.firebase.crashlytics.FirebaseCrashlytics;
import com.google.firebase.messaging.FirebaseMessagingService;
import com.google.firebase.messaging.RemoteMessage;

import cl.figonzal.lastquakechile.R;
import cl.figonzal.lastquakechile.services.SharedPrefService;
import timber.log.Timber;

public class FirebaseQuakeNotificationService extends FirebaseMessagingService {

    private FirebaseCrashlytics crashlytics;
    private QuakesNotification quakesNotification;

    @Override
    public void onCreate() {
        super.onCreate();

        crashlytics = FirebaseCrashlytics.getInstance();
        quakesNotification = new QuakesNotification(getApplicationContext(), new SharedPrefService(getApplicationContext()));
    }

    @Override
    public void onMessageReceived(@NonNull RemoteMessage remoteMessage) {
        super.onMessageReceived(remoteMessage);

        Timber.tag(getString(R.string.TAG_FIREBASE_MESSAGE)).i("From: %s", remoteMessage.getFrom());

        //Si es notificacion con datos de sismos
        if (remoteMessage.getData().size() > 0) {

            Timber.tag(getString(R.string.TAG_FIREBASE_MESSAGE)).i("Message data payload: %s", remoteMessage.getData());
            crashlytics.setCustomKey(getString(R.string.FIREBASE_MESSAGE_DATA_STATUS), true);

            quakesNotification.setRemoteMsg(remoteMessage);
            quakesNotification.showNotification();
        }

        //Si es notificacion desde consola FCM
        if (remoteMessage.getNotification() != null) {

            Timber.tag(getString(R.string.TAG_FIREBASE_MESSAGE)).i("Message notification: " + remoteMessage.getNotification().getTitle() + " - " + remoteMessage.getNotification().getBody());
            crashlytics.setCustomKey(getString(R.string.FIREBASE_MESSAGE_NOTIFICATION_STATUS), true);

            quakesNotification.showNotificationGeneric();
        }
    }

    /**
     * Funcion que muestra notificacion desde PHP
     */
    @Override
    public void onNewToken(@NonNull String s) {
        super.onNewToken(s);

        Timber.tag(getString(R.string.TAG_FIREBASE_TOKEN)).i("Refreshed Token: %s", s);
        crashlytics.setUserId(s);
    }

    @Override
    public void onDeletedMessages() {
        super.onDeletedMessages();
    }
}

File: app/src/main/java/cl/figonzal/lastquakechile/services/notifications/FirebaseQuakeNotificationService.kt
package cl.figonzal.lastquakechile.services.notifications

import cl.figonzal.lastquakechile.R
import cl.figonzal.lastquakechile.core.utils.SharedPrefUtil
import com.google.firebase.crashlytics.FirebaseCrashlytics
import com.google.firebase.messaging.FirebaseMessagingService
import com.google.firebase.messaging.RemoteMessage
import timber.log.Timber

class FirebaseQuakeNotificationService : FirebaseMessagingService() {

    private lateinit var crashlytics: FirebaseCrashlytics
    private var quakesNotification: QuakesNotification? = null

    //Esta clase no puede tener contructor
    //REF: https://firebase.google.com/docs/cloud-messaging?hl=es
    override fun onCreate() {
        super.onCreate()
        crashlytics = FirebaseCrashlytics.getInstance()
        quakesNotification = QuakesNotification(
            applicationContext, SharedPrefUtil(
                applicationContext
            )
        )
    }

    override fun onMessageReceived(remoteMessage: RemoteMessage) {
        super.onMessageReceived(remoteMessage)
        Timber.i("From: %s", remoteMessage.from)

        //Si es notificacion con datos de sismos
        if (remoteMessage.data.isNotEmpty()) {

            Timber.i("Message data payload: %s", remoteMessage.data)
            crashlytics.setCustomKey(getString(R.string.FIREBASE_MESSAGE_DATA_STATUS), true)

            quakesNotification?.showNotification(remoteMessage)
        }

        //Si es notificacion desde consola FCM
        if (remoteMessage.notification != null) {
            Timber.i(
                "Message notification: " + remoteMessage.notification!!
                    .title + " - " + remoteMessage.notification!!.body
            )
            crashlytics.setCustomKey(
                getString(R.string.FIREBASE_MESSAGE_NOTIFICATION_STATUS),
                true
            )
            quakesNotification?.showNotificationGeneric(remoteMessage)
        }
    }

    /**
     * Funcion que muestra notificacion desde PHP
     */
    override fun onNewToken(s: String) {
        super.onNewToken(s)
        Timber.i("Refresh token: %s", s)
        crashlytics.setUserId(s)
    }
}